src/utils/collection.rs

```diff
@@ -198,6 +198,42 @@ impl BtfCollection {
         Ok(ids)
     }
 
+    /// Find a list of BTF ids using a regex describing their name as a key.
+    /// Matching ids can be found in multiple underlying BTF, thus this function
+    /// returns a list of tuples containing each a reference to `NamedBtf`
+    /// (representing the BTF where a match was found) and the id. Further
+    /// lookups must be done using the `Btf` object contained in the linked
+    /// `NamedBtf` one.
+    #[cfg(feature = "regex")]
+    pub fn resolve_ids_by_regex(&self, re: &regex::Regex) -> Result<Vec<(&NamedBtf, u32)>> {
+        let mut ids = Vec::new();
+
+        let mut base_ids = match self.base.btf.resolve_ids_by_regex(re) {
+            Ok(base_ids) => base_ids,
+            _ => Vec::new(), // Id not found in base.
+        };
+
+        for split in self.split.iter() {
+            if let Ok(mut mod_ids) = split.btf.resolve_ids_by_regex(re) {
+                // Add ids from the split BTF if not already in the base lookup,
+                // as each split BTF will also lookup in the base BTF.
+                mod_ids
+                    .drain(..)
+                    .filter(|i| !base_ids.iter().any(|b| b == i))
+                    .for_each(|i| ids.push((split, i)));
+            }
+        }
+
+        // Now add ids found in the base BTF.
+        base_ids.drain(..).for_each(|i| ids.push((&self.base, i)));
+
+        if ids.is_empty() {
+            bail!("No id linked to regex {re}");
+        }
```

Comment:
I know all other  API functions do the same but this got me thinking that returning an error if we didn't find any id/name is not the best API.
There are many things that can go wrong and the user cannot easily distinguish what happened (without matching on a string which is is bad).

For me (as a user), it makes more sense for a function that "searches" to reaturn an error if the search was not even done because something went wrong (e.g: the regexp I passed is uncomprehensible) and return an empty Vector (or Iterator) if the search did not find anything. I can then maybe change the search parameter or do something about it. Hiding empty searches as errors forces me to also error out.

Of course this is just a general comment, maybe a follow up PR can handle all cases when the current nested PRs get merged.

Reply:
That makes perfect sense. I had a quick look at all those examples and IMO they can be easily converted. I noted many of those helpers can't return an error (except for the `empty()` case).

This makes me thinking:
- Should we keep returning a `Result` in such cases. It might be good future proof, but really for many except for having no results this can't fail (either there is a match in our internal storage or not)?
- Should we return an empty `Vec` in case of no result, or embed it into an `Option`; making things more explicit? This also maps to how iterator work.

Right now if I were to decide I would remove the `Result` and use an `Option` instead (or a `Result<Option<>>` in case it can fail, eg. regex). But no strong opinion.

(Yes this would break the API, but nothing prevents us from releasing v2 at some point).

Reply:
`Option` or `Result<Option>` makes sense to me as well. If we then move to an interator, the return type is the same at least.